src/flow/src/adapter.rs

```diff
@@ -14,19 +14,668 @@
+            flow_err_collectors: Default::default(),
+            src_send_buf_lens: Default::default(),
+            tick_manager,
+            node_id,
+        }
+    }
+
+    /// Create a flownode manager with one worker
+    pub fn new_with_worker<'s>(
+        node_id: Option<u32>,
+        query_engine: Arc<dyn QueryEngine>,
+        table_meta: TableMetadataManagerRef,
+    ) -> (Self, Worker<'s>) {
+        let mut zelf = Self::new(node_id, query_engine, table_meta);
+        let (handle, worker) = create_worker();
+        zelf.add_worker_handle(handle);
+        (zelf, worker)
+    }
+
+    /// add a worker handler to manager, meaning this corresponding worker is under it's manage
+    pub fn add_worker_handle(&mut self, handle: WorkerHandle) {
+        self.worker_handles.push(Mutex::new(handle));
+    }
+}
+
+#[derive(Debug)]
+pub enum DiffRequest {
+    Insert(Vec<(Row, repr::Timestamp)>),
+    Delete(Vec<(Row, repr::Timestamp)>),
+}
+
+/// iterate through the diff row and form continuous diff row with same diff type
+pub fn diff_row_to_request(rows: Vec<DiffRow>) -> Vec<DiffRequest> {
+    let mut reqs = Vec::new();
+    for (row, ts, diff) in rows {
+        let last = reqs.last_mut();
+        match (last, diff) {
+            (Some(DiffRequest::Insert(rows)), 1) => {
+                rows.push((row, ts));
+            }
+            (Some(DiffRequest::Insert(_)), -1) => reqs.push(DiffRequest::Delete(vec![(row, ts)])),
+            (Some(DiffRequest::Delete(rows)), -1) => {
+                rows.push((row, ts));
+            }
+            (Some(DiffRequest::Delete(_)), 1) => reqs.push(DiffRequest::Insert(vec![(row, ts)])),
+            (None, 1) => reqs.push(DiffRequest::Insert(vec![(row, ts)])),
+            (None, -1) => reqs.push(DiffRequest::Delete(vec![(row, ts)])),
+            _ => {}
+        }
+    }
+    reqs
+}
+
+/// This impl block contains methods to send writeback requests to frontend
+impl FlownodeManager {
+    /// TODO(discord9): merge all same type of diff row into one requests
+    ///
+    /// Return the number of requests it made
+    pub async fn send_writeback_requests(&self) -> Result<usize, Error> {
+        let all_reqs = self.generate_writeback_request().await;
+        if all_reqs.is_empty() || all_reqs.iter().all(|v| v.1.is_empty()) {
+            return Ok(0);
+        }
+        let mut req_cnt = 0;
+        for (table_name, reqs) in all_reqs {
+            if reqs.is_empty() {
+                continue;
+            }
+            let (catalog, schema) = (table_name[0].clone(), table_name[1].clone());
+            let ctx = Arc::new(QueryContext::with(&catalog, &schema));
+            // TODO(discord9): instead of auto build table from request schema, actually build table
+            // before `create flow` to be able to assign pk and ts etc.
+            let (primary_keys, schema) = if let Some(table_id) = self
+                .table_info_source
```

Comment:
Can we use `CatalogManager` directly? It is good at caching for `Table`. cc @WenyXu

Reply:
> Question: can we use `CatalogManager` directly? It is good at caching for `Table`. cc @WenyXu

There is some need to query table from given `TableId` (i.e. when query source table's schema when accepting `RegionInsertRequest` wit h`TableId` from frontend) but CatalogManager doesn't have similiar methods?